Net: Defer reading until listeners could be added

fastest963 opened the pull request:
Defer reading until user-land has a chance to add listeners. This
allows the TLS wrapper to listen for _tlsError and trigger a
clientError event if the socket already has data that could trigger.

Fixes #1114

fastest963 commented:
This also fixes https://github.com/joyent/node/issues/9355

chrisdickinson commented on lib/_tls_wrap.js:
```diff
@@ -247,7 +247,19 @@ function TLSSocket(socket, options) {
+  var self = this;
+  process.nextTick(function() {
+    // Socket already has some buffered data - emulate receiving it
+    if (socket && socket._readableState && socket._readableState.length) {
```
I know this has just been moved, but we really need to standardize an API for querying the size of a stream's buffer.

chrisdickinson commented on lib/_tls_wrap.js:
```diff
@@ -383,13 +395,6 @@ TLSSocket.prototype._init = function(socket, wrap) {
-    while ((buf = socket.read()) !== null)
-      ssl.receive(buf);
-  }
-
```
Why move this out of `_init` into the constructor?

fastest963 replied:
I moved it to put it in the `process.nextTick` that I added to the constructor. If we want to just do the `nextTick` here AND in the constructor around `read(0)`, I'm fine doing it in both places, just let me know.

brendanashworth replied:
Why do it in both places, wouldn't doing it once in `_init` complete the same job?

fastest963 replied:
Unless I'm missing something, I need to defer the `ssl.receive(buf);` in order to fix the issue I'm trying to fix with this PR. If I defer that, I need to ALSO defer the `this.read(0)` in the constructor or it will start to read data off the socket before the existing data was read in. Please correct me if I'm wrong.

chrisdickinson replied:
Ah! That totally makes sense now. Thanks for clarifying – I missed that `.read(0)` was in the constructor.

chrisdickinson commented on lib/_tls_wrap.js:
```diff
@@ -247,7 +247,19 @@ function TLSSocket(socket, options) {
+    }
+
+    self.read(0);
+  });
```
possibly of note: `process.nextTick` can take arguments now, so we can avoid having to close over `self` (passing it as an argument instead) and allowing us to put the function at module scope.

fastest963 replied:
Awesome, thanks for the tip. I'll add that once we come to a resolution with your previous comment (doing nextTick in constructor and _init or only once).

fastest963 replied:
That would make it more difficult to port back to Node though right, since they didn't implement the `nextTick` args, yet? Do we care about that?

chrisdickinson replied:
It should be fine to make use of `nextTick` args in this PR. It's likely that the Node patches will be merged on top of iojs when the convergence WG starts up (so we're not worried about porting back patches for now.)

brendanashworth commented on test/parallel/test-tls-delayed-attach-error.js:
```diff
@@ -0,0 +1,50 @@
+    });
+
+    s.on('close', function() {
+      ended++;
```
Is this necessary? Since you don't `.unref()` the server it wouldn't close the process naturally and never have `process.on('exit')` called, right?

fastest963 replied:
That seems right. I was mostly just copying `test-tls-delayed-attach.js` which did this.

indutny replied:
Ok...

fastest963 commented:
@chrisdickinson Updated to send along `this` as an arg to `nextTick` and added a check for `if (!self._handle)`. I also removed `ended` from the tests that @brendanashworth brought up. Since we're not using `unref` we don't need to check that it ended since the process won't die until it has.

Thanks guys for taking a look at this so quickly.

chrisdickinson commented on lib/_tls_wrap.js:
```diff
@@ -247,7 +247,22 @@ function TLSSocket(socket, options) {
+    }
+
+    self.read(0);
+  }, this);
```
could you name this anonymous function and pull it out of the constructor?

fastest963 replied:
Yup! Updated and moved into a named function above the constructor called `initread`

fastest963 commented:
@chrisdickinson I moved the anonymous function to above the constructor. Everything else look good? Thanks again!

chrisdickinson commented:
LGTM! If @indutny does not object, I'll merge this tomorrow.

indutny commented on lib/_tls_wrap.js:
```diff
@@ -199,6 +199,20 @@ function onocspresponse(resp) {
   this.emit('OCSPResponse', resp);
 }
 
+function initread(tls, wrapped) {
```
Camel case, please.

fastest963 replied:
Gotcha. I was trying to follow the function above `onnewsession` which was all lowercase so I thought that was the standard :/

indutny commented on test/parallel/test-tls-delayed-attach.js:
```diff
@@ -12,7 +12,6 @@ var net = require('net');
 
 var sent = 'hello world';
 var received = '';
-var ended = 0;
```
Why?

fastest963 replied:
See https://github.com/iojs/io.js/pull/1496#discussion_r28851104

fastest963 commented:
@indutny I renamed to `initRead`. If you want me to keep the `ended` var and checks i the tests, I can add those back, just let me know. Thanks

fastest963 commented:
@indutny would you mind taking another look after the changes I made? Thanks!

fastest963 commented:
Any updates @indutny @chrisdickinson ?

indutny commented:
LGTM, @chrisdickinson please land!

brendanashworth commented on test/parallel/test-tls-delayed-attach-error.js:
```diff
@@ -0,0 +1,45 @@
+var common = require('../common');
```
A `'use strict';` now needs to be added here with the new linter.

fastest963 commented:
@brendanashworth updated with `'use strict';` and linter seems to like it and tests still pass.

@chrisdickinson good to go?

brendanashworth commented:
CI: https://jenkins-iojs.nodesource.com/view/iojs/job/iojs+any-pr+multi/10/

brendanashworth commented:
hmm... the linter job failed. @fastest963 could you rebase your remote branch onto master and push to GH (preferably with a `make lint` beforehand)? It might need an update.

fastest963 commented:
@brendanashworth It looks like it failed because of:
```
[linter] $ /bin/sh -xe /tmp/hudson2733538223575635564.sh
+ NODE=/usr/local/bin/iojs make lint
gmake[1]: Entering directory '/usr/home/iojs/build/workspace/iojs+linter/nodes/linter'
./iojs tools/eslint/bin/eslint.js src lib test --reset --quiet
gmake[1]: ./iojs: Command not found
Makefile:378: recipe for target 'jslint' failed
gmake[1]: *** [jslint] Error 127
gmake[1]: Leaving directory '/usr/home/iojs/build/workspace/iojs+linter/nodes/linter'
*** Error code 2
```

brendanashworth commented:
I don't work on build but I think you need the recent updates to the
Makefile to lint on the CI. <surprise!> Could you rebase, then I'll try to
run it again.

Sorry about this, its a new CI setup. I'm still getting used to it.

On Monday, June 15, 2015, James Hartig <notifications@github.com> wrote:

> @brendanashworth <https://github.com/brendanashworth> It looks like it
> failed because of:
>
> [linter] $ /bin/sh -xe /tmp/hudson2733538223575635564.sh
> + NODE=/usr/local/bin/iojs make lint
> gmake[1]: Entering directory '/usr/home/iojs/build/workspace/iojs+linter/nodes/linter'
> ./iojs tools/eslint/bin/eslint.js src lib test --reset --quiet
> gmake[1]: ./iojs: Command not found
> Makefile:378: recipe for target 'jslint' failed
> gmake[1]: *** [jslint] Error 127
> gmake[1]: Leaving directory '/usr/home/iojs/build/workspace/iojs+linter/nodes/linter'
> *** Error code 2
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nodejs/io.js/pull/1496#issuecomment-112249838>.
>

fastest963 commented:
No problem! I just rebased and pushed, want to try again? Thanks!

brendanashworth commented:
perfecto: https://jenkins-iojs.nodesource.com/job/iojs+any-pr+multi/14/

brendanashworth commented:
Thanks for the pull request! Landed in https://github.com/nodejs/io.js/commit/061342a50075a23e04465e0ac2f33124ab56ea32.
